siliconcompiler/design.py

```diff
@@ -0,0 +1,535 @@
+
+        Args:
+            fileset (str or list[str]): Fileset(s) to export.
+            filename (str or Path): Output file name.
+            fileformat (str, optional): Export format.
+
+        Inferred from file extension if not given.
+
+        """
+
+        if filename is None:
+            raise ValueError("write_fileset() filename cannot be None")
+
+        if not isinstance(fileset, list):
+            fileset = [fileset]
+
+        # file extension lookup
+        if not fileformat:
+            formats = {}
+            formats['f'] = 'flist'
+            fileformat = formats[Path(filename).suffix.strip('.')]
+
+        if fileformat == "flist":
+            # TODO: handle dependency tree
+            # TODO: add source info for comments to flist.
+            with open(filename, "w") as f:
+                for i in fileset:
+                    if not isinstance(i, str):
+                        raise ValueError("fileset key must be a string")
+                    for j in ['idir', 'define', 'file']:
+                        if j == 'idir':
+                            vals = self.get('fileset', i, 'idir')
+                            cmd = "+incdir+"
+                        elif j == 'define':
+                            vals = self.get('fileset', i, 'define')
+                            cmd = "+define+"
+                        else:
+                            vals = self.get('fileset', i, 'file', 'verilog')
+                            cmd = ""
+                        if vals:
+                            for item in vals:
+                                f.write(f"{cmd}{item}\n")
+        else:
+            raise ValueError(f"{fileformat} is not supported")
+
+    ################################################
+    def read_fileset(self, fileset: str, filename: str, fileformat=None) -> None:
+        """Imports filesets from a standard formatted text file.
+
+        Currently supports Verilog `flist` format only.
+        Intended to support other formats in the future.
+
+        Args:
+            fileset (str or list[str]): Fileset(s) to import.
+            filename (str or Path): Output file name.
+            fileformat (str, optional): Export format.
+                Inferred from file extension if not given.
+
+        """
+
+        if filename is None:
+            raise ValueError("read_fileset() filename cannot be None")
+
+        if not fileformat:
+            formats = {}
+            formats['f'] = 'flist'
+            fileformat = formats[Path(filename).suffix.strip('.')]
+
+        if fileformat == "flist":
+            raise RuntimeError("read_fileset is not implemented yet")
+        else:
+            raise ValueError(f"{fileformat} is not supported")
+
+    ################################################
+    # Helper Functions
+    ################################################
+    def _set(self, fileset, option, value, typelist=None):
```

Comment:
this should be `__set` to avoid exposing this to users (this is purely an internal function)

Reply:
done